File: tests/cpp/test_streaming.cpp
// (C) Copyright 2024- ECMWF and individual contributors.
//
// This software is licensed under the terms of the Apache Licence Version 2.0
// which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
// In applying this licence, ECMWF does not waive the privileges and immunities
// granted to it by virtue of its status as an intergovernmental organisation nor
// does it submit to any jurisdiction.

// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 ECMWF
//
// Tests for tensogram::streaming_encoder.

#include <gtest/gtest.h>
#include <tensogram.hpp>
#include "test_helpers.hpp"

#include <cstdio>
#include <cstring>
#include <vector>

using test_helpers::TempFile;

// ---------------------------------------------------------------------------
// Helper: descriptor JSON
// ---------------------------------------------------------------------------

namespace {

/// Per-object descriptor JSON for a 1-D float32 array.
std::string f32_descriptor(std::size_t count) {
    return R"({"type":"ndarray","ndim":1,"shape":[)" +
           std::to_string(count) +
           R"(],"strides":[4],"dtype":"float32","byte_order":"little","encoding":"none","filter":"none","compression":"none"})";
}

} // anonymous namespace

// ---------------------------------------------------------------------------
// Create, write objects, finish, then open and decode
// ---------------------------------------------------------------------------

TEST(StreamingTest, CreateWriteFinishDecode) {
    TempFile tmp;

    std::string meta_json = R"({"version":2})";
    {
        tensogram::streaming_encoder enc(tmp.path, meta_json);

        std::vector<float> values = {1.0f, 2.0f, 3.0f};
        enc.write_object(
            f32_descriptor(values.size()),
            reinterpret_cast<const std::uint8_t*>(values.data()),
            values.size() * sizeof(float));

        enc.finish();
    }

    // Open the file and decode
    auto f = tensogram::file::open(tmp.path);
    EXPECT_EQ(f.message_count(), 1u);

    auto msg = f.decode_message(0);
    EXPECT_EQ(msg.num_objects(), 1u);

    auto obj = msg.object(0);
    EXPECT_EQ(obj.dtype_string(), "float32");
    EXPECT_EQ(obj.element_count<float>(), 3u);
    const float* p = obj.data_as<float>();
    EXPECT_FLOAT_EQ(p[0], 1.0f);
    EXPECT_FLOAT_EQ(p[1], 2.0f);
    EXPECT_FLOAT_EQ(p[2], 3.0f);
}

// ---------------------------------------------------------------------------
// object_count increments after each write
// ---------------------------------------------------------------------------

TEST(StreamingTest, ObjectCountIncrements) {
    TempFile tmp;

    std::string meta_json = R"({"version":2})";
    tensogram::streaming_encoder enc(tmp.path, meta_json);

    EXPECT_EQ(enc.object_count(), 0u);

    std::vector<float> v1 = {1.0f};
    enc.write_object(
        f32_descriptor(v1.size()),
        reinterpret_cast<const std::uint8_t*>(v1.data()),
        v1.size() * sizeof(float));
    EXPECT_EQ(enc.object_count(), 1u);

    std::vector<float> v2 = {2.0f, 3.0f};
    enc.write_object(
        f32_descriptor(v2.size()),
        reinterpret_cast<const std::uint8_t*>(v2.data()),
        v2.size() * sizeof(float));
    EXPECT_EQ(enc.object_count(), 2u);

    enc.finish();
}

// ---------------------------------------------------------------------------
// Multi-object streaming write
// ---------------------------------------------------------------------------

TEST(StreamingTest, MultiObjectStreaming) {
    TempFile tmp;

    std::string meta_json = R"({"version":2})";
    tensogram::streaming_encoder enc(tmp.path, meta_json);

    for (int i = 0; i < 5; ++i) {
        std::vector<float> values = {static_cast<float>(i * 10)};
        enc.write_object(
            f32_descriptor(values.size()),
            reinterpret_cast<const std::uint8_t*>(values.data()),
            values.size() * sizeof(float));
    }
    enc.finish();

    auto f = tensogram::file::open(tmp.path);
    auto msg = f.decode_message(0);
    ASSERT_EQ(msg.num_objects(), 5u);

    for (std::size_t i = 0; i < 5; ++i) {
        auto obj = msg.object(i);
        EXPECT_FLOAT_EQ(obj.data_as<float>()[0],
                        static_cast<float>(i * 10));
    }
}

// ---------------------------------------------------------------------------
// Read back the streamed file and verify data
// ---------------------------------------------------------------------------

TEST(StreamingTest, ReadBackVerifyData) {
    TempFile tmp;

    std::string meta_json = R"({"version":2})";
    {
        tensogram::streaming_encoder enc(tmp.path, meta_json);

        std::vector<float> v1 = {42.0f, 43.0f};
        std::vector<float> v2 = {99.0f};
        enc.write_object(
            f32_descriptor(v1.size()),
            reinterpret_cast<const std::uint8_t*>(v1.data()),
            v1.size() * sizeof(float));
        enc.write_object(
            f32_descriptor(v2.size()),
            reinterpret_cast<const std::uint8_t*>(v2.data()),
            v2.size() * sizeof(float));
        enc.finish();
    }

    // Read raw bytes, then decode
    auto f = tensogram::file::open(tmp.path);
    auto raw = f.read_message(0);
    ASSERT_FALSE(raw.empty());

    auto msg = tensogram::decode(raw.data(), raw.size());
    ASSERT_EQ(msg.num_objects(), 2u);

    auto obj0 = msg.object(0);
    EXPECT_EQ(obj0.element_count<float>(), 2u);
    EXPECT_FLOAT_EQ(obj0.data_as<float>()[0], 42.0f);

    auto obj1 = msg.object(1);
    EXPECT_EQ(obj1.element_count<float>(), 1u);
    EXPECT_FLOAT_EQ(obj1.data_as<float>()[0], 99.0f);
}

// ---------------------------------------------------------------------------
// Streaming encoder RAII (destructor frees without crash)
// ---------------------------------------------------------------------------

TEST(StreamingTest, StreamingEncoderRAII) {
    TempFile tmp;

    // Create and immediately destroy WITHOUT calling finish()
    {
        std::string meta_json = R"({"version":2})";
        tensogram::streaming_encoder enc(tmp.path, meta_json);

        std::vector<float> values = {1.0f};
        enc.write_object(
            f32_descriptor(values.size()),
            reinterpret_cast<const std::uint8_t*>(values.data()),
            values.size() * sizeof(float));
        // enc goes out of scope without finish() — destructor should free
    }

    // If we get here without crash, RAII is working
    SUCCEED();
}

// ---------------------------------------------------------------------------
// Streaming with hash algorithm
// ---------------------------------------------------------------------------

TEST(StreamingTest, StreamingWithHash) {
    TempFile tmp;

    std::string meta_json = R"({"version":2})";
    tensogram::encode_options opts;
    opts.hash_algo = "xxh3";

    {
        tensogram::streaming_encoder enc(tmp.path, meta_json, opts);

        std::vector<float> values = {1.0f, 2.0f};
        enc.write_object(
            f32_descriptor(values.size()),
            reinterpret_cast<const std::uint8_t*>(values.data()),
            values.size() * sizeof(float));
        enc.finish();
    }

    auto f = tensogram::file::open(tmp.path);
    tensogram::decode_options dec_opts;
    dec_opts.verify_hash = true;
    auto msg = f.decode_message(0, dec_opts);
    EXPECT_EQ(msg.num_objects(), 1u);

    auto obj = msg.object(0);
    EXPECT_TRUE(obj.has_hash());
    EXPECT_EQ(obj.hash_type(), "xxh3");
    EXPECT_FALSE(obj.hash_value().empty());
}

// ---------------------------------------------------------------------------
// Preceder metadata: write_preceder + write_object round-trip
// ---------------------------------------------------------------------------

TEST(StreamingTest, PrecederRoundTrip) {
    TempFile tmp;

    std::string meta_json = R"({"version":2})";
    {
        tensogram::streaming_encoder enc(tmp.path, meta_json);

        // Write preceder for first object
        enc.write_preceder(R"({"mars":{"param":"2t"},"units":"K"})");

        std::vector<float> values = {1.0f, 2.0f, 3.0f};
        enc.write_object(
            f32_descriptor(values.size()),
            reinterpret_cast<const std::uint8_t*>(values.data()),
            values.size() * sizeof(float));

        enc.finish();
    }

    // Open and decode — verify data survives the preceder + data object path.
    // (Preceder metadata merges into base[0] on the Rust side;
    // the C++ metadata API accesses per-object keys via base entries.)
    auto f = tensogram::file::open(tmp.path);
    auto msg = f.decode_message(0);
    EXPECT_EQ(msg.num_objects(), 1u);

    auto obj = msg.object(0);
    EXPECT_EQ(obj.dtype_string(), "float32");
    EXPECT_EQ(obj.element_count<float>(), 3u);
    EXPECT_FLOAT_EQ(obj.data_as<float>()[0], 1.0f);
    EXPECT_FLOAT_EQ(obj.data_as<float>()[1], 2.0f);
    EXPECT_FLOAT_EQ(obj.data_as<float>()[2], 3.0f);
}

// ---------------------------------------------------------------------------
// Preceder: mixed objects (some with, some without)
// ---------------------------------------------------------------------------

TEST(StreamingTest, PrecederMixedObjects) {
    TempFile tmp;

    std::string meta_json = R"({"version":2})";
    {
        tensogram::streaming_encoder enc(tmp.path, meta_json);

        // Object 0: with preceder
        enc.write_preceder(R"({"note":"obj0"})");
        std::vector<float> v0 = {10.0f};
        enc.write_object(
            f32_descriptor(v0.size()),
            reinterpret_cast<const std::uint8_t*>(v0.data()),
            v0.size() * sizeof(float));

        // Object 1: no preceder
        std::vector<float> v1 = {20.0f};
        enc.write_object(
            f32_descriptor(v1.size()),
            reinterpret_cast<const std::uint8_t*>(v1.data()),
            v1.size() * sizeof(float));

        enc.finish();
    }

    auto f = tensogram::file::open(tmp.path);
    auto msg = f.decode_message(0);
    ASSERT_EQ(msg.num_objects(), 2u);
    EXPECT_FLOAT_EQ(msg.object(0).data_as<float>()[0], 10.0f);
    EXPECT_FLOAT_EQ(msg.object(1).data_as<float>()[0], 20.0f);
}

// ---------------------------------------------------------------------------
// Preceder: consecutive write_preceder without write_object throws
// ---------------------------------------------------------------------------

TEST(StreamingTest, PrecederDoubleWriteThrows) {
    TempFile tmp;

    std::string meta_json = R"({"version":2})";
    tensogram::streaming_encoder enc(tmp.path, meta_json);

    enc.write_preceder(R"({})");
    EXPECT_THROW(enc.write_preceder(R"({})"), tensogram::framing_error);
}

// ---------------------------------------------------------------------------
// Preceder: non-object JSON throws
// ---------------------------------------------------------------------------

TEST(StreamingTest, PrecederNonObjectJsonThrows) {
    TempFile tmp;

    std::string meta_json = R"({"version":2})";
    tensogram::streaming_encoder enc(tmp.path, meta_json);

    EXPECT_THROW(enc.write_preceder(R"([1,2,3])"), tensogram::metadata_error);
}

File: fortran/test/test_streaming.f90
! (C) Copyright 2026- ECMWF and individual contributors.
!
! This software is licensed under the terms of the Apache Licence Version 2.0
! which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
! In applying this licence, ECMWF does not waive the privileges and immunities
! granted to it by virtue of its status as an intergovernmental organisation nor
! does it submit to any jurisdiction.

!> Streaming encoder: write a single multi-object message progressively (one
!> data object at a time, mixed dtype/rank, with lossless compression on one),
!> finish, then reopen and decode all objects and assert a bit-identical
!> round-trip. PLAN_FORTRAN.md streaming milestone.
program test_streaming
   use, intrinsic :: iso_c_binding
   use tensogram
   implicit none

   character(len=*), parameter :: path = 'test_streaming_tmp.tgm'
   type(tensogram_streaming_encoder) :: enc
   type(tensogram_file)              :: f
   type(tensogram_message)           :: msg
   integer(c_int) :: err
   integer :: i, j, k, n, ios

   real(c_float)      :: a1(3, 4)
   real(c_double)     :: a2(5)
   integer(c_int32_t) :: a3(2, 2, 2)
   real(c_float),      allocatable :: o1(:,:)
   real(c_double),     allocatable :: o2(:)
   integer(c_int32_t), allocatable :: o3(:,:,:)

   do j = 1, 4; do i = 1, 3; a1(i, j) = real(i, c_float) + 0.1_c_float * real(j, c_float); end do; end do
   do i = 1, 5; a2(i) = real(i, c_double) * 2.5_c_double; end do
   do k = 1, 2; do j = 1, 2; do i = 1, 2; a3(i, j, k) = i * 100 + j * 10 + k; end do; end do; end do

   ! Stream three objects of different dtype/rank, one at a time.
   call tensogram_streaming_encoder_create(path, enc, err); call assert(err == TGM_ERROR_OK, 'create')
   call tensogram_streaming_encoder_write(enc, a1, err, compression='zstd')  ! lossless
   call assert(err == TGM_ERROR_OK, 'write obj1 (zstd)')
   call assert(tensogram_streaming_encoder_count(enc) == 1, 'count == 1')
   call tensogram_streaming_encoder_write(enc, a2, err); call assert(err == TGM_ERROR_OK, 'write obj2')
   call tensogram_streaming_encoder_write(enc, a3, err); call assert(err == TGM_ERROR_OK, 'write obj3')
   call assert(tensogram_streaming_encoder_count(enc) == 3, 'count == 3')
   call tensogram_streaming_encoder_finish(enc, err);    call assert(err == TGM_ERROR_OK, 'finish')
   call enc%free()

   ! Reopen the finished file: one message, three objects, all round-trip.
   call tensogram_file_open(path, f, err);            call assert(err == TGM_ERROR_OK, 'open')
   call tensogram_file_message_count(f, n, err);      call assert(err == TGM_ERROR_OK, 'count err')
   call assert(n == 1, 'one message')
   call tensogram_file_decode_message(f, 1, msg, err); call assert(err == TGM_ERROR_OK, 'decode')
   call assert(tensogram_num_objects(msg) == 3, 'three objects')

   call tensogram_to_array(msg, 1, o1, err); call assert(err == TGM_ERROR_OK, 'to_array 1')
   call assert(all(shape(o1) == [3, 4]), 'obj1 shape')
   call assert(bytes_eq(transfer(o1, [0_c_int8_t]), transfer(a1, [0_c_int8_t])), 'obj1 values (zstd lossless)')

   call tensogram_to_array(msg, 2, o2, err); call assert(err == TGM_ERROR_OK, 'to_array 2')
   call assert(size(o2) == 5, 'obj2 size')
   call assert(bytes_eq(transfer(o2, [0_c_int8_t]), transfer(a2, [0_c_int8_t])), 'obj2 values')

   call tensogram_to_array(msg, 3, o3, err); call assert(err == TGM_ERROR_OK, 'to_array 3')
   call assert(all(shape(o3) == [2, 2, 2]), 'obj3 shape')
   call assert(all(o3 == a3), 'obj3 values')

   call f%close()
   open(newunit=ios, file=path, status='old', iostat=err)
   if (err == 0) close(ios, status='delete')

   print '(a)', 'test_streaming: PASS'

contains

   subroutine assert(cond, what)
      logical,          intent(in) :: cond
      character(len=*), intent(in) :: what
      if (.not. cond) then
         print '(a,a)', 'test_streaming: FAIL: ', what
         error stop 1
      end if
   end subroutine assert

   logical function bytes_eq(x, y)
      integer(c_int8_t), intent(in) :: x(:), y(:)
      bytes_eq = size(x) == size(y)
      if (bytes_eq) bytes_eq = all(x == y)
   end function bytes_eq

end program test_streaming
